File: interpretor/bfi.c
#include <stdio.h>
#include <stdlib.h>

signed char c;
unsigned char mem[65536];
unsigned char * ptr = mem;
unsigned char * startPtr = mem;
FILE * file;

void block() {
    long int blockPos = ftell(file);

    while (!feof(file))
    {   
        c = fgetc(file);

        switch (c) {
            case '[':
                block();
                break;
            case ']':
                if (*ptr) fseek(file, blockPos, SEEK_SET);
                else return;
                break;
            case '+':
                ++(*ptr);
                break;
            case '-':
                if ((*ptr) > 0) --(*ptr);
                break;
            case '>':
                ++ptr;
                break;
            case '<':
                if (startPtr < ptr) --ptr;
                break;
            case ',': 
                scanf("%c", ptr);
                if (*ptr == '\n') *ptr = 0;
                break;
            case '.':
                printf("%c", *ptr);
        }
    }
}

int main(int argc, char ** argv) {
    if (argv[1] == NULL) {
        printf("File name was not specified\n");
        return 0;
    } 

    file = fopen(argv[1], "rb+");

    if (file == NULL) {
        printf("No such file named '%s' \n", argv[1]);
        return 0;
    }

    block(file);
    fclose(file);
    printf("\n");

    return 0;
} 

File: interpretor/bfi.py
import sys

file = None
mem = [0] * 65536
ptr = 0
inputStack = ''

def block():
    global ptr, mem, inputStack
    pos = file.tell()
    
    while True:
        c = file.read(1)

        if c == '[':
            block()
        elif c == ']':
            if mem[ptr]:
                file.seek(pos)
            else:
                return
        elif c == '+':
            mem[ptr] += 1
        elif c =='-':
            if mem[ptr] > 0:
                mem[ptr] -= 1
        elif c == '>':
            if ptr < 65535:
                ptr += 1
        elif c == '<':
            if ptr > 0:
                ptr -= 1
        elif c == '.':
            print(chr(mem[ptr]), end = '')
        elif c == ',':
            if not inputStack:
                inputStack = input('')
            mem[ptr] = ord(inputStack[0])
            inputStack = inputStack[1:]

def main():
    global file

    if (len(sys.argv) < 2):
        print('File name was not specified')
        return

    try:
        file = open(sys.argv[1], 'r')
    except:
        print(f"No such file named '{sys.argv[1]}'")
        return

    block()
    file.close()
    print()

if __name__ == '__main__':
    main()